tasmota/xsns_65_hdc1080.ino

```diff
@@ -0,0 +1,349 @@
+
+  hdc_humidity = HDC1080_RH_MULT * (float) (rh_data);
+
+  if (hdc_humidity > 100) { hdc_humidity = 100.0; }
+  if (hdc_humidity < 0) { hdc_humidity = 0.01; }
+
+  ConvertHumidity(hdc_humidity);  // Set global humidity
+
+  hdc_valid = SENSOR_MAX_MISS;
+
+  return true;
+}
+
+/**
+ * Performs the detection of the HTC1080 sensor.
+ * 
+ */
+
+void HdcDetect(void) {
+  if (I2cActive(HDC1080_ADDR)) { 
+    AddLog_P2(LOG_LEVEL_DEBUG, PSTR("HdcDetect: Address = 0x%02X already in use."), HDC1080_ADDR);
+
+    return; 
+  }
+
+  hdc_manufacturer_id = HdcReadManufacturerId();
+  hdc_device_id = HdcReadDeviceId();
+
+  AddLog_P2(LOG_LEVEL_DEBUG, PSTR("HdcDetect: detected device with manufacturerId = 0x%04X and deviceId = 0x%04X"), hdc_manufacturer_id, hdc_device_id);
+
+  if (hdc_device_id == HDC1080_DEV_ID) {
+    HdcInit();
+    I2cSetActiveFound(HDC1080_ADDR, hdc_type_name);
+  }
+}
+
+/**
+ * As the name suggests, this function is called every second
+ * for performing driver related logic.
+ * 
+ */
+void HdcEverySecond(void) {
+  if (uptime &1) {  // Every 2 seconds
+    if (!HdcTriggerRead()) {
+      AddLogMissed((char*) hdc_type_name, hdc_valid);
+    }
+  }
+}
+
+/**
+ * Tasmota boilerplate for presenting the sensor data in the web UI, JSON for 
+ * the MQTT messages, and so on.
+ * 
+ */
+void HdcShow(bool json) {
+  if (hdc_valid) {
+    char temperature[33];
+
+    dtostrfd(hdc_temperature, Settings.flag2.temperature_resolution, temperature);
+    char humidity[33];
+    dtostrfd(hdc_humidity, Settings.flag2.humidity_resolution, humidity);
+
+    if (json) {
+      ResponseAppend_P(JSON_SNS_TEMPHUM, hdc_type_name, temperature, humidity);
+#ifdef USE_DOMOTICZ
+      if (0 == tele_period) {
+        DomoticzTempHumSensor(temperature, humidity);
+      }
+#endif  // USE_DOMOTICZ
+#ifdef USE_KNX
+      if (0 == tele_period) {
+        KnxSensor(KNX_TEMPERATURE, hdc_temperature);
+        KnxSensor(KNX_HUMIDITY, hdc_humidity);
+      }
+#endif  // USE_KNX
+#ifdef USE_WEBSERVER
+    } else {
+      WSContentSend_PD(HTTP_SNS_TEMP, hdc_type_name, temperature, TempUnit());
+      WSContentSend_PD(HTTP_SNS_HUM, hdc_type_name, humidity);
+#endif  // USE_WEBSERVER
+    }
+  }
+}
+
+/*********************************************************************************************\
+ * Interface
+\*********************************************************************************************/
+
+bool Xsns65(uint8_t function)
+{
+  if (!I2cEnabled(XI2C_45)) { 
+    AddLog_P(LOG_LEVEL_DEBUG, PSTR("Xsns65: I2C driver not enabled for this device."));
+
+    return false; 
+  }
+
+  bool result = false;
+
+  if (FUNC_INIT == function) {
+    HdcDetect();
+  }
+  else if (hdc_device_id) {
+    switch (function) {
+      case FUNC_EVERY_50_MSECOND:
+        if(is_reading && TimeReached(timer)) {
+          if(!HdcRead()) {
+            AddLogMissed((char*) hdc_type_name, hdc_valid);
+          }
+          timer = millis() + HDC1080_CONV_TIME;
```

Comment:
This line is not needed. The timer must be set just after the command.

Reply:
@s-hadinger  which line? When it ticks I need to update the timer variable (not a timer in itself, just a the value of the next instant when something needs to be done) to the closest instant after which I expect data to be available in the sensor.

Reply:
But ok @s-hadinger ok I see the point. The time it takes to finish the read from the previous iteration will condition how much time is left as the timer is armed again.